File: EvalVisitor.py
from antlr4 import *
from LabeledExprVisitor import LabeledExprVisitor
from LabeledExprParser import LabeledExprParser
import math

class EvalVisitor(LabeledExprVisitor):
    def __init__(self):
        super().__init__()
        self.memory = {}
        self.degrees = False  # False = rad, True = deg

    # ID '=' expr NEWLINE
    def visitAssign(self, ctx:LabeledExprParser.AssignContext):
        varname = ctx.ID().getText()
        value = self.visit(ctx.expr())
        self.memory[varname] = value
        return value

    # expr NEWLINE
    def visitPrintExpr(self, ctx:LabeledExprParser.PrintExprContext):
        value = self.visit(ctx.expr())
        print(self._format_number(value))
        return 0.0

    # NEWLINE
    def visitBlank(self, ctx:LabeledExprParser.BlankContext):
        return 0.0

    # DEG NEWLINE
    def visitSetDeg(self, ctx:LabeledExprParser.SetDegContext):
        self.degrees = True
        return 0.0

    # RAD NEWLINE
    def visitSetRad(self, ctx:LabeledExprParser.SetRadContext):
        self.degrees = False
        return 0.0

    # NUMBER
    def visitNumber(self, ctx:LabeledExprParser.NumberContext):
        return float(ctx.NUMBER().getText())

    # ID
    def visitId(self, ctx:LabeledExprParser.IdContext):
        name = ctx.ID().getText()
        return self.memory.get(name, 0.0)

    # expr op=('*'|'/') expr
    def visitMulDiv(self, ctx:LabeledExprParser.MulDivContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.MUL:
            return left * right
        else:
            return left / right

    # expr op=('+'|'-') expr
    def visitAddSub(self, ctx:LabeledExprParser.AddSubContext):
        left = self.visit(ctx.expr(0))
        right = self.visit(ctx.expr(1))
        if ctx.op.type == LabeledExprParser.ADD:
            return left + right
        else:
            return left - right

    # SUB expr
    def visitNegate(self, ctx:LabeledExprParser.NegateContext):
        return -self.visit(ctx.expr())

    # expr FACT
    def visitFactorial(self, ctx:LabeledExprParser.FactorialContext):
        x = self.visit(ctx.expr())
        if (abs(x - round(x)) > 1e-12) or (x < 0):
            raise RuntimeError("Factorial sólo admite enteros no negativos.")
        return float(self._factorial(int(round(x))))

    # func '(' expr ')'
    def visitFunc(self, ctx:LabeledExprParser.FuncContext):
        x = self.visit(ctx.expr())
        f = ctx.func().getText().lower()
        if f == 'sin':
            return math.sin(self._to_radians_if_needed(x))
        elif f == 'cos':
            return math.cos(self._to_radians_if_needed(x))
        elif f == 'tan':
            return math.tan(self._to_radians_if_needed(x))
        elif f == 'sqrt':
            if x < 0:
                raise RuntimeError("Sqrt(x): x no puede ser negativo.")
            return math.sqrt(x)
        elif f == 'ln':
            if x <= 0:
                raise RuntimeError("Ln(x): x debe ser > 0.")
            return math.log(x)
        elif f == 'log':
            if x <= 0:
                raise RuntimeError("Log(x): x debe ser > 0.")
            return math.log10(x)
        else:
            raise RuntimeError(f"Función no soportada: {f}")

    # '(' expr ')'
    def visitParens(self, ctx:LabeledExprParser.ParensContext):
        return self.visit(ctx.expr())

    # ---------- utilidades ----------
    def _factorial(self, n:int) -> int:
        res = 1
        for i in range(2, n+1):
            res *= i
        return res

    def _to_radians_if_needed(self, x:float) -> float:
        return math.radians(x) if self.degrees else x

    def _format_number(self, v:float) -> str:
        if abs(v - round(v)) < 1e-12:
            return str(int(round(v)))
        return str(v)

File: EvalVisitor.java
import java.util.HashMap;
import java.util.Map;

public class EvalVisitor extends LabeledExprBaseVisitor<Double> {
    /** Memoria para variables */
    Map<String, Double> memory = new HashMap<>();

    /** Modo ángulo: true = grados, false = radianes (por defecto radianes) */
    private boolean degrees = false;

    /** ID '=' expr NEWLINE */
    @Override
    public Double visitAssign(LabeledExprParser.AssignContext ctx) {
        String id = ctx.ID().getText();
        double value = visit(ctx.expr());
        memory.put(id, value);
        return value;
    }

    /** expr NEWLINE */
    @Override
    public Double visitPrintExpr(LabeledExprParser.PrintExprContext ctx) {
        Double value = visit(ctx.expr());
        System.out.println(formatNumber(value));
        return 0.0;
    }

    /** NEWLINE (blank) */
    @Override
    public Double visitBlank(LabeledExprParser.BlankContext ctx) {
        return 0.0;
    }

    /** DEG NEWLINE */
    @Override
    public Double visitSetDeg(LabeledExprParser.SetDegContext ctx) {
        degrees = true;
        return 0.0;
    }

    /** RAD NEWLINE */
    @Override
    public Double visitSetRad(LabeledExprParser.SetRadContext ctx) {
        degrees = false;
        return 0.0;
    }

    /** NUMBER */
    @Override
    public Double visitNumber(LabeledExprParser.NumberContext ctx) {
        return Double.valueOf(ctx.NUMBER().getText());
    }

    /** ID */
    @Override
    public Double visitId(LabeledExprParser.IdContext ctx) {
        String id = ctx.ID().getText();
        return memory.getOrDefault(id, 0.0);
    }

    /** expr op=('*'|'/') expr */
    @Override
    public Double visitMulDiv(LabeledExprParser.MulDivContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        if (ctx.op.getType() == LabeledExprParser.MUL) return left * right;
        return left / right; // DIV
    }

    /** expr op=('+'|'-') expr */
    @Override
    public Double visitAddSub(LabeledExprParser.AddSubContext ctx) {
        double left = visit(ctx.expr(0));
        double right = visit(ctx.expr(1));
        if (ctx.op.getType() == LabeledExprParser.ADD) return left + right;
        return left - right; // SUB
    }

    /** SUB expr (unario) */
    @Override
    public Double visitNegate(LabeledExprParser.NegateContext ctx) {
        return -visit(ctx.expr());
    }

    /** expr FACT (factorial) */
    @Override
    public Double visitFactorial(LabeledExprParser.FactorialContext ctx) {
        double x = visit(ctx.expr());
        if (!isWholeNumber(x) || x < 0) {
            throw new RuntimeException("Factorial sólo admite enteros no negativos.");
        }
        return (double) factorial((long)Math.round(x));
    }

    /** func '(' expr ')' */
    @Override
    public Double visitFunc(LabeledExprParser.FuncContext ctx) {
        double x = visit(ctx.expr());
        String f = ctx.func().getText().toLowerCase();

        switch (f) {
            case "sin": return Math.sin(toRadiansIfNeeded(x));
            case "cos": return Math.cos(toRadiansIfNeeded(x));
            case "tan": return Math.tan(toRadiansIfNeeded(x));
            case "sqrt":
                if (x < 0) throw new RuntimeException("Sqrt(x): x no puede ser negativo.");
                return Math.sqrt(x);
            case "ln":
                if (x <= 0) throw new RuntimeException("Ln(x): x debe ser > 0.");
                return Math.log(x);
            case "log":
                if (x <= 0) throw new RuntimeException("Log(x): x debe ser > 0.");
                return Math.log10(x);
            default:
                throw new RuntimeException("Función no soportada: " + f);
        }
    }

    /** '(' expr ')' */
    @Override
    public Double visitParens(LabeledExprParser.ParensContext ctx) {
        return visit(ctx.expr());
    }

    // utilidades 

    private boolean isWholeNumber(double v) {
        return Math.abs(v - Math.rint(v)) < 1e-12;
    }

    private long factorial(long n) {
        long res = 1;
        for (long i = 2; i <= n; i++) res *= i;
        return res;
    }

    private double toRadiansIfNeeded(double x) {
        return degrees ? Math.toRadians(x) : x;
    }

    private String formatNumber(double v) {
        // Evita "2.0" cuando es entero
        if (isWholeNumber(v)) return String.valueOf((long)Math.rint(v));
        return String.valueOf(v);
    }
}
